core/trace/src/main/scala-2/org/typelevel/otel4s/trace/TracerMacro.scala

```diff
@@ -104,7 +103,8 @@ private[otel4s] trait TracerMacro[F[_]] {
   def resourceSpan[A](name: String, attributes: Attribute[_]*)(
       resource: Resource[F, A]
   ): Resource[F, Span.Res[F, A]] =
-    macro TracerMacro.resourceSpan[F, A]
+     macro TracerMacro.resourceSpan[F, A]
+   */
```

Comment:
This function is the biggest loss in local semantics: we can only hug the acquire and release, not the use.

Reply:
With changes from https://github.com/typelevel/otel4s/pull/86, we can do the trick:

```scala
val resource: Resource[IO, Unit] = Resource.make(IO.sleep(50.millis))(_ => IO.sleep(100.millis))

tracer.span("resource").wrapResource(resource).use { _ =>
  Work[IO].doWork
}
```
<img width="1076" alt="image" src="https://user-images.githubusercontent.com/6395483/216267596-27839882-3af2-4104-8ab2-e6d56ab9054c.png">

Reply:
That branch still has this "stateful" definition of `createScope`:

https://github.com/typelevel/otel4s/blob/48ecdf4bde2d3b5ea48237302521d3c9e15af619/java/trace/src/main/scala/org/typelevel/otel4s/java/trace/TraceScope.scala#L84-L87

I think that's where we run into the problems with `interruptScope`, vs. this "local" definition:

https://github.com/typelevel/otel4s/blob/2a206466bd2104b4c5b028de2ef4dadc62b08353/java/trace/src/main/scala/org/typelevel/otel4s/java/trace/TraceScope.scala#L83-L89

I don't yet see how to make `wrapResource` work with the local semantics.  If we do, we get nice traces like the above _and_ avoid the corruption via common fs2 combinators, but I'm still not convinced it's possible.

Reply:
Hmm.  Your #86 works even in an FS2 `interruptScope`.  🤔 

```scala
  def run: IO[Unit] =
    tracerResource.use { implicit tracer =>
      val resource: Resource[IO, Unit] = Resource.make(IO.sleep(50.millis))(_ => IO.sleep(100.millis))

      def stream(name: String) =
        Stream.eval(tracer.spanBuilder(name).wrapResource(resource).start.surround(
          Work[IO].doWork
        ))

      tracer.span("root").surround(
        (stream("uninterrupted") ++ stream("interrupted").interruptScope).compile.drain
      )
    }
```

<img width="1391" alt="Jaeger console screenshot showing use still sibling of acquire and release in both interrupted and uninterrupted subtraces" src="https://user-images.githubusercontent.com/142698/216347016-a98d84d6-3c3a-4f11-88bd-0859cfcd09eb.png">

Reply:
What is the expected output (a.k.a. `interruptScope` break tracing)? Incorrect parent for `stream("interrupted").interruptScope`?

Reply:
Ah, but this does not.  The second `DoWork` escapes from the `interrupted` span:

```scala
    tracerResource.use { implicit tracer =>
      val resource: Resource[IO, Unit] = Resource.make(IO.sleep(50.millis))(_ => IO.sleep(100.millis))

      def stream(name: String) =
        Stream.resource(tracer.spanBuilder(name).wrapResource(resource).start)
          .evalMap(_ => Work[IO].doWork)

      tracer.span("root").surround(
        (stream("uninterrupted") ++ stream("interrupted").interruptScope).compile.drain
      )
    }
```

<img width="1391" alt="Screenshot 2023-02-02 at 9 14 33 AM" src="https://user-images.githubusercontent.com/142698/216348615-dbe2aa35-d61a-44fa-a7c1-ec029c59f2ff.png">

(Apologies for the lack of alt text ... I need to figure out how to dump these spans to a text tree...)

Reply:
You can print spans as a tree in tests 

https://github.com/typelevel/otel4s/blob/main/java/trace/src/test/scala/org/typelevel/otel4s/java/trace/TracerSuite.scala#L470

The format is `{span_name} {start} -> {end}`
```
resource-span 1000000000 -> 1675000000 =>
  acquire 1000000000 -> 1025000000 =>
    acquire-inner 1000000000 -> 1025000000
  use 1025000000 -> 1375000000 =>
    body-1 1025000000 -> 1125000000
    body-2 1125000000 -> 1325000000
    body-3 1325000000 -> 1375000000
  release 1375000000 -> 1675000000
```

Reply:
Once we merge #105, we will not be able to do `fs2.Stream.resource(tracer.spanBuilder(name).wrapResource(resource).start)`, since there is no resource exposed.

Here is an example with both #105 and #86:
https://github.com/typelevel/otel4s/pull/107/files#diff-1864b2af85047d6d543da3ec001b74204471c7489f78a01fcb5db5f502480b81R58-R66

Reply:
Ooh, neat, this is the first time both good ideas have been on one branch.  I think we'd need a wrapStream, but I also think it might work.  Let me see if I can break this one!  :laughing: